### PR TITLE
Only send samesite cookies

### PR DESCRIPTION
**Pull request by rullzer**
This makes the last remaining two cookies lax. The session cookie
itself. And the session password as well (on php 7.3 that is). Samesite
cookies are the best cookies!

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

### CODE REVIEW

File: lib/private/Session/CryptoWrapper.php [hunk 1]
```diff
@@ -86,7 +86,23 @@ public function __construct(IConfig $config,
...
+					setcookie(self::COOKIE_NAME, $this->passphrase, 0, $webRoot, '', $secureCookie, true);
+				} else {
+					setcookie(
+						self::COOKIE_NAME,
+						$this->passphrase,
+						[
+							'expires' => 0,
+							'path' => $webRoot,
+							'domain' => '',
+							' secure' => $secureCookie,
```
Comment (ChristophWurst) [inline lib/private/Session/CryptoWrapper.php]:
```suggestion
							'secure' => $secureCookie,
```

?

**Review by ChristophWurst — CHANGES_REQUESTED** [review]
Clicked around, shared some files. All fine :+1:

File: lib/private/Session/CryptoWrapper.php (hunk 1 above)
Comment (rullzer) [reply lib/private/Session/CryptoWrapper.php]: it was late. lets leave it at that.

**Comment by rullzer** [comment]
So... for some reason this does :boom: when mixed. So not this only happens on php7.3 and later.
We should probably test in older browsers. But then I guess this is good to go!

File: lib/private/Session/CryptoWrapper.php [hunk 2]
```diff
@@ -86,7 +86,23 @@ public function __construct(IConfig $config,
 				if($webRoot === '') {
 					$webRoot = '/';
 				}
-				setcookie(self::COOKIE_NAME, $this->passphrase, 0, $webRoot, '', $secureCookie, true);
+
+				if (version_compare(PHP_VERSION, '7.3') === -1) {
```
Comment (kesselb) [inline lib/private/Session/CryptoWrapper.php]: Please consider `PHP_VERSION_ID` ;)

Comment (kesselb) [reply lib/private/Session/CryptoWrapper.php]:
```suggestion
				if (\PHP_VERSION_ID < 70300) {
```

File: lib/private/Session/Internal.php [hunk 3]
```diff
@@ -210,4 +210,12 @@ private function invoke(string $functionName, array $parameters = [], bool $sile
 			$this->trapError($e->getCode(), $e->getMessage());
 		}
 	}
+
+	private function startSession() {
+		if (version_compare(PHP_VERSION, '7.3') === -1) {
```
Comment (kesselb) [inline lib/private/Session/Internal.php]:
```suggestion
		if (\PHP_VERSION_ID < 70300) {
```

File: lib/private/Session/CryptoWrapper.php (hunk 2 above)
Comment (kesselb) [inline lib/private/Session/CryptoWrapper.php]:
```suggestion
				if (\PHP_VERSION_ID < 70300) {
```

Comment (kesselb) [reply lib/private/Session/CryptoWrapper.php]: It wasn't me? ;)

Comment (skjnldsv) [reply lib/private/Session/CryptoWrapper.php]: Shaggy?

Comment (kesselb) [reply lib/private/Session/CryptoWrapper.php]: I recall that was related to have three of those suggested changes. I'm certain that I pressed the comment button only once ;)

**Review by nickvergessen — APPROVED** [review]

**Review by ChristophWurst — APPROVED** [review]

**Comment by peacepenguin** [comment]
This change breaks SAML auth in many cases with PHP 7.3 and above.

During a Nextcloud-SAML login, the user attempts to goto the nextcloud frontpage to login, nextcloud sets up the session cookies right away. Then redirects the user to login via SAML.

After the SAML login, there is a POST request with the SAML authorization back to nextcloud: 
 https://<nextcloud fqdn>/index.php/apps/user_saml/saml/acs

With these cookies set to LAX, the POST doesn't contain the needed session cookies, and the auth fails entirely.

If we drop the PHP version to 7.2, the cookie doesn't get set to LAX, and the SAML response POST will contain the needed session cookie, and login via SAML is successful.

This phenomenon is documented here, https://php.watch/articles/PHP-Samesite-cookies :

> When a cookie is marked samesite=Lax, that cookie will not be passed for any cross-domain requests unless it's a regular link that navigates user to the target site. Other requests methods (such as **POST** and PUT) and XHR requests will not contain this cookie.

I can confirm this is the source of the issue two different ways: drop the server php version to 7.2, issue gone; makes sense since this change code only sets the flag if php is 7.3 or above.

Also in chrome hitting nextcloud 19 with php 7.3 or 7.4, after the failed SAML login attempt from this issue, you can manually modify the session cookies in the F12 dev tools to remove the LAX flag from the two session cookies, then retrying login succeeds. 

A workaround is setting up the SAML IdP server on the same subdomain as the nextcloud instance:

sso.**mydomain.net**
nextcloud.**mydomain.net**

In this scenario, the bug doesn't manifest, since they are both part of the same subdomain, LAX allows the cookie to be sent even in a POST.

Unfortunately changing the hosted FQDN of the Identity provider to match your nextcloud instance may not be an option, or feasible.

Perhaps an override for this cookie change can be put in place so SAML auth app users can continue to function as before?

**Pull request merged**